src/terminal/message.rs

```diff
@@ -1,9 +1,50 @@
 use super::emoji;
+use once_cell::sync::OnceCell;
 
 use billboard::{Billboard, BorderColor, BorderStyle};
+use serde::{Deserialize, Serialize};
+
+#[derive(PartialEq)]
+pub enum OutputType {
+    Json,
+    Human,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct PublishOutput {
+    pub success: Option<bool>,
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    pub urls: Vec<String>,
+}
+
+static OUTPUT_TYPE: OnceCell<OutputType> = OnceCell::new();
+
+pub fn set_output_type(typ: OutputType) {
+    let _ = OUTPUT_TYPE.set(typ);
+}
+
+pub fn is_json_output() -> bool {
+    OUTPUT_TYPE.get() == Some(&OutputType::Json)
+}
+
+// Always goes to stdout
+pub fn jsonout<T>(value: &T)
+where
+    T: ?Sized + Serialize,
+{
+    println!("{}", &serde_json::to_string(value).unwrap());
+}
 
 fn message(msg: &str) {
-    println!("{}", msg);
+    match OUTPUT_TYPE.get() {
+        Some(OutputType::Json) => {
+            eprintln!("{}", msg);
```

Comment:
are these backwards? i think json should go to stdout and human readable should be stderr, no?

Reply:
> json should go to stdout and human readable should be stderr, no?
Yes- IFF json output flag is given. I made the assumption messages are for human readable output. It made sense to me to make a json type here. I'm a rust novice so let me know if there is a better way. 
`wrangler build | publish --output json`

Wrangler supports structured output for core commands `build` and `publish`. `JSON` string goes to stdout, stderr displays user output.

js
```
nat@cf:~/examplejs$wrangler build
💁  JavaScript project found. Skipping unnecessary build!
nat@cf:~/examplejs$wrangler build --output json
{"success":true,"name":"examplejs"}
nat@davidson:~/examplejs$ ../wrangler/target/debug/wrangler publish --output json `returns no output, need to fix.

```
Webpack `build` and `publish`
```
nat@davidson:~/demo-site$ ../wrangler/target/debug/wrangler build --output json
{"success":true,"name":"demo-site"}
nat@davidson:~/demo-site$ ../wrangler/target/debug/wrangler publish --output json
{"success":true,"name":"demo-site","urls":["https://demo-site.nats.workers.dev"]}

```
Next steps: extend to other worker types